### PR TITLE
3.0 iproto uri params

### LINKED ISSUE #3945: new format of the `iproto.*` config options.
**Related dev. issue(s):** https://github.com/tarantool/tarantool/pull/9385
**Related doc. issue(s):** https://github.com/tarantool/doc/issues/3665, https://github.com/tarantool/enterprise_doc/issues/225

**Product:** Tarantool
**Since:** 3.0
**Root document:**
- https://www.tarantool.io/en/doc/latest/concepts/configuration/ (including env vars)
- https://www.tarantool.io/en/doc/latest/reference/configuration/configuration_reference/
- https://www.tarantool.io/en/doc/latest/how-to/replication/
- All config samples: https://github.com/tarantool/doc/tree/latest/doc/code_snippets/snippets

**SME:** @ totktonada, @ ImeevMA

# Details

The config options `iproto.listen`, `iproto.advertise.peer` and
`iproto.advertise.sharding` now have a new format.

The `iproto.listen` config option now has the following format:
```
iproto:
  listen:
    - uri: <string>
      params:
        transport: <'plain' or 'ssl'>
        ssl_ca_file: <string>
        ssl_cert_file: <string>
        ssl_ciphers: <string>
        ssl_key_file: <string>
        ssl_password: <string>
        ssl_password_file: <string>
    - uri: <string>
      params:
        transport: <'plain' or 'ssl'>
        ssl_ca_file: <string>
        ssl_cert_file: <string>
        ssl_ciphers: <string>
        ssl_key_file: <string>
        ssl_password: <string>
        ssl_password_file: <string>
...
```
The `iproto.listen` config option is now an array, and its elements are
`records` that can only have a required `uri` field and an optional
`params` field.

The `uri` field is of type `string` and must be a single valid URI. The
URI cannot contain parameters, login and password.

The `params` field is a `record` with the fields `transport`,
`ssl_ca_file`, `ssl_cert_file`, `ssl_ciphers`, `ssl_key_file`,
`ssl_password`, `ssl_password_file`. The `transport` field can have one
of two values: `plain` or `ssl`, the default is `plain`. If `transport`
is set to `plain`, no additional fields are required. When `transport`
is set to `ssl`, the fields `ssl_key_file` and `ssl_cert_file` are
required, and the remaining fields are optional.

The `iproto.advertise.peer` and `iproto.advertise.sharding` now have the
following format:
```
iproto:
  advertise:
    <'peer' or 'sharding'>:
      uri: <string>
      login: <string>
      password: <string>
      params:
        transport: <'plain' or 'ssl'>
        ssl_ca_file: <string>
        ssl_cert_file: <string>
        ssl_ciphers: <string>
        ssl_key_file: <string>
        ssl_password: <string>
        ssl_password_file: <string>
```

The `iproto.advertise.peer` and `iproto.advertise.sharding` are now
`records`. All of their fields are optional, however, if the `params`
field can be set only if the `uri` field is set and the `password` field
can be set only if the `login` field is set.

The `uri` field is of type `string` and must be a single valid URI.
The URI cannot contain parameters, login and password. If the `uri` and
`params` fields are missing, the appropriate `iproto.listen` element is
selected. An element is eligible if, after parsing its `url` field using
`require('uri').parse()`, the `ipv4` result field is not equal to
`'0.0.0.0'`, the `ipv6` result field is not equal to `'::'`, and the
`service` result field is not equal to `'0'`.

The `login` and `password` fields are of type `string`.

For description of the `params` field see description of
`iproto.listen`.
Requested by @ImeevMA in https://github.com/tarantool/tarantool/commit/1247f47adc9b991e7362cd8c74d3f5a5bdea8832.

### PR DESCRIPTION
**Pull request by andreyaksenov**
1) Created a runnable sample with SSL enabled for all instances:
    - https://github.com/tarantool/doc/tree/3.0-iproto-uri-params/doc/code_snippets/snippets/replication/instances.enabled/ssl
    
   The config from this sample is included to the `<uri>.params.transport` option description.

2) Documented all options in `<uri>.params.*`:
     - https://docs.d.tarantool.io/en/doc/3.0-iproto-uri-params/reference/configuration/configuration_reference/#uri-params

     Options descriptions are taken from [Traffic encryption](https://www.tarantool.io/en/doc/latest/enterprise/security/#traffic-encryption) with some cosmetic changes.

> [!NOTE]  
> The `Traffic encryption` section will be updated to the new config and moved to the [Configuration](https://www.tarantool.io/en/doc/latest/concepts/configuration/) section when working on the https://github.com/tarantool/doc/issues/3665 issue.

3) Updated some existing descriptions:

    - `iproto_advertise.<peer_or_sharding>.uri`: https://docs.d.tarantool.io/en/doc/3.0-iproto-uri-params/reference/configuration/configuration_reference/#iproto-advertise-peer-or-sharding
    - `iproto.listen`: https://docs.d.tarantool.io/en/doc/3.0-iproto-uri-params/reference/configuration/configuration_reference/#configuration-reference-iproto-misc

4) Updated the `net.box` sample (includes code from the new SSL sample):
     - https://docs.d.tarantool.io/en/doc/3.0-iproto-uri-params/enterprise/security/#configuration

### CODE REVIEW

**Review by ImeevMA — APPROVED** [review]
LGTM!

File: doc/reference/configuration/configuration_reference.rst [hunk 1]
```diff
@@ -619,7 +619,8 @@ iproto.advertise.<peer_or_sharding>.*
 
 .. confval:: iproto_advertise.<peer_or_sharding>.uri
 
-    An URI used to advertise the current instance.
+    (Optional) An URI used to advertise the current instance.
```
Comment (p7nov) [inline doc/reference/configuration/configuration_reference.rst]:
```suggestion
    (Optional) A URI used to advertise the current instance.
```

File: doc/reference/configuration/configuration_reference.rst [hunk 2]
```diff
@@ -634,7 +635,8 @@ iproto.advertise.<peer_or_sharding>.*
 
 .. confval:: iproto_advertise.<peer_or_sharding>.login
 
-    A username that should be used to connect to the current instance.
+    (Optional) A username that should be used to connect to the current instance.
```
Comment (p7nov) [inline doc/reference/configuration/configuration_reference.rst]:
```suggestion
    (Optional) A username used for connecting to the current instance.
```
Feel free to ignore this edit if it changes the meaning.

File: doc/reference/configuration/configuration_reference.rst [hunk 3]
```diff
@@ -645,8 +647,8 @@ iproto.advertise.<peer_or_sharding>.*
 
 .. confval:: iproto_advertise.<peer_or_sharding>.password
 
-    A password for the specified user.
-    If a password is missing, it is taken from :ref:`credentials <configuration_reference_credentials>` for the specified username.
+    (Optional) A password for the specified user.
+    If ``login`` is specified but a password is missing, it is taken from :ref:`credentials <configuration_reference_credentials>` for this user.
```
Comment (p7nov) [inline doc/reference/configuration/configuration_reference.rst]:
```suggestion
    If a ``login`` is specified but a password is missing, it is taken from the user's :ref:`credentials <configuration_reference_credentials>`.
```

File: doc/reference/configuration/configuration_reference.rst [hunk 4]
```diff
@@ -775,24 +776,84 @@ iproto.*
...
+    **Example**
+
+    The example below demonstrates how to enable traffic encryption for connections between replica set peers.
+    The following parameters are specified for each instance:
+
+    -   ``ssl_ca_file``: a path to a trusted certificate authorities (CA) file.
+    -   ``ssl_cert_file``: a path to an SSL certificate file.
+    -   ``ssl_key_file``: a path to a private SSL key file.
+    -   ``ssl_password`` (``instance001``): a password for an encrypted private SSL key.
+        For ``instance002`` and ``instance003``, passwords for encrypted SSL keys are provided in ``ssl_password_file``.
```
Comment (p7nov) [inline doc/reference/configuration/configuration_reference.rst]: Maybe make `ssl_password_file` a separate list item for better visibility?

**Review by p7nov — APPROVED** [review]
LGTM, just some minor comments.

File: doc/reference/configuration/configuration_reference.rst (hunk 1 above)
Comment (andreyaksenov) [reply doc/reference/configuration/configuration_reference.rst]: Thanks, also fixed in other places

File: doc/reference/configuration/configuration_reference.rst (hunk 2 above)
Comment (andreyaksenov) [reply doc/reference/configuration/configuration_reference.rst]: Changed to `A username used to connect to the current instance.`

**Pull request merged**